File: backend/agent/agent.js
#!/usr/bin/env node

require("dotenv").config({ quiet: true });

const WebSocket = require("ws");
const Docker = require("dockerode");
const dns = require("dns");
const fs = require("fs");
const os = require("os");
const path = require("path");

const WS_URL = "wss://dockmon.onrender.com/agent";

const docker =
  process.platform === "win32"
    ? new Docker({ socketPath: "//./pipe/docker_engine" })
    : new Docker({ socketPath: "/var/run/docker.sock" });

const CONFIG_PATH = path.join(os.homedir(), ".dockmon", "config.json");

const getToken = () => {
  try {
    if (!fs.existsSync(CONFIG_PATH)) return null;
    const raw = fs.readFileSync(CONFIG_PATH, "utf8");
    if (!raw) return null;
    const data = JSON.parse(raw);
    return data.deviceToken || null;
  } catch (err) {
    console.error("⚠️ Failed to read config:", err.message);
    return null;
  }
};

const deviceToken = getToken();

if (!deviceToken) {
  console.error("❌ No DEVICE_TOKEN found.");
  console.error("👉 Run: dockmon-agent login");
  process.exit(1);
}

/*
---------------------------------------
STATE
---------------------------------------
*/

let ws = null;
let reconnectTimer = null;
let reconnectDelay = 1000;
const MAX_DELAY = 5000;
let shuttingDown = false;
let isConnecting = false;
let lastConnectedAt = 0;
const CONNECT_COOLDOWN = 4000;

/*
---------------------------------------
LOG
---------------------------------------
*/

const log = (msg, meta) => {
  const suffix = meta ? ` ${JSON.stringify(meta)}` : "";
  console.log(`[${new Date().toISOString()}] ${msg}${suffix}`);
};

/*
---------------------------------------
UTILS
---------------------------------------
*/

const sendJson = (payload) => {
  if (ws && ws.readyState === WebSocket.OPEN) {
    ws.send(JSON.stringify(payload));
  }
};

const formatContainer = (c) => ({
  id: c.Id,
  names: c.Names,
  image: c.Image,
  state: c.State,
  status: c.Status,
});

const clearReconnectTimer = () => {
  if (reconnectTimer) {
    clearTimeout(reconnectTimer);
    reconnectTimer = null;
  }
};

const isConnected = () => ws && ws.readyState === WebSocket.OPEN;

/*
---------------------------------------
HELPERS
---------------------------------------
*/

const getStats = (container) =>
  new Promise((resolve, reject) => {
    container.stats({ stream: false }, (err, stats) => {
      if (err) reject(err);
      else resolve(stats);
    });
  });

/*
---------------------------------------
HANDLERS
---------------------------------------
*/

const handlers = {
  async list_containers() {
    const containers = await docker.listContainers({ all: true });
    return containers.map(formatContainer);
  },
  async start_container({ container_id }) {
    await docker.getContainer(container_id).start();
    return { success: true };
  },
  async stop_container({ container_id }) {
    await docker.getContainer(container_id).stop();
    return { success: true };
  },
  async restart_container({ container_id }) {
    await docker.getContainer(container_id).restart();
    return { success: true };
  },
  async remove_container({ container_id }) {
    await docker.getContainer(container_id).remove({ force: true });
    return { success: true };
  },
  async logs({ container_id }) {
    const container = docker.getContainer(container_id);
    const buffer = await container.logs({
      stdout: true,
      stderr: true,
      timestamps: true,
      tail: 500,
    });
    if (!buffer) return "No logs";
    return buffer.toString("utf8") || "No logs";
  },
  async stats({ container_id }) {
    const stats = await getStats(docker.getContainer(container_id));
    const cpuDelta =
      stats.cpu_stats.cpu_usage.total_usage -
      stats.precpu_stats.cpu_usage.total_usage;
    const systemDelta =
      stats.cpu_stats.system_cpu_usage - stats.precpu_stats.system_cpu_usage;
    const cpuCount =
      stats.cpu_stats.online_cpus ||
      stats.cpu_stats.cpu_usage.percpu_usage.length;
    const cpu =
      systemDelta > 0 && cpuDelta > 0
        ? (cpuDelta / systemDelta) * cpuCount * 100
        : 0;
    return {
      cpu_percent: Number(cpu.toFixed(2)),
      memory_usage: stats.memory_stats.usage,
      memory_limit: stats.memory_stats.limit,
    };
  },
  async list_images() {
    return await docker.listImages();
  },
  async pull_image({ image }) {
    return new Promise((resolve, reject) => {
      docker.pull(image, (err, stream) => {
        if (err) return reject(err);
        docker.modem.followProgress(stream, (err) => {
          if (err) reject(err);
          else resolve({ success: true, image });
        });
      });
    });
  },
  async run_container({ image, name }) {
    const container = await docker.createContainer({ Image: image, name });
    await container.start();
    return { success: true, container_id: container.id };
  },
};

/*
---------------------------------------
COMMAND HANDLER
---------------------------------------
*/

const handleCommand = async (msg) => {
  const { request_id, command, payload } = msg;
  log("📥 Command", { command });
  if (!handlers[command]) {
    log("⚠️ Unsupported", { command });
    return sendJson({ type: "response", request_id, data: { error: "Unsupported command" } });
  }
  try {
    const result = await handlers[command](payload || {});
    sendJson({ type: "response", request_id, data: result });
  } catch (err) {
    sendJson({ type: "response", request_id, data: { error: err.message } });
  }
};

/*
---------------------------------------
CONNECT
---------------------------------------
*/

const connect = () => {
  if (shuttingDown) return;
  if (isConnected()) return;
  if (isConnecting) return;

  clearReconnectTimer();
  isConnecting = true;

  if (ws) {
    try { ws.terminate(); } catch {}
    ws = null;
  }

  log("🔌 Connecting...");
  ws = new WebSocket(WS_URL);

  ws.on("open", () => {
    isConnecting = false;
    reconnectDelay = 1000;
    lastConnectedAt = Date.now();
    log("✅ Connected");
    sendJson({ type: "register", device_token: deviceToken });
  });

  ws.on("message", async (raw) => {
    try {
      const msg = JSON.parse(raw.toString());
      if (msg.type === "command") await handleCommand(msg);
    } catch (e) {
      log("❌ Parse error", { error: e.message });
    }
  });

  ws.on("close", () => {
    isConnecting = false;
    const timeSinceConnect = Date.now() - lastConnectedAt;
    if (timeSinceConnect < CONNECT_COOLDOWN) {
      clearReconnectTimer();
      log("🔁 Brief disconnect — waiting 3s");
      reconnectTimer = setTimeout(() => {
        reconnectTimer = null;
        connect();
      }, 3000);
    } else {
      scheduleReconnect();
    }
  });

  ws.on("error", (err) => {
    isConnecting = false;
    log("⚠️ WS Error", { error: err.message });
  });
};

/*
---------------------------------------
IMMEDIATE RECONNECT
---------------------------------------
*/

const immediateReconnect = (reason) => {
  if (shuttingDown) return;
  if (isConnected()) return;
  if (isConnecting) return;

  const timeSinceConnect = Date.now() - lastConnectedAt;
  if (timeSinceConnect < CONNECT_COOLDOWN) return;

  log(`⚡ ${reason} — reconnecting immediately`);
  clearReconnectTimer();
  reconnectDelay = 1000;
  connect();
};

/*
---------------------------------------
SCHEDULED RECONNECT (with backoff)
---------------------------------------
*/

const scheduleReconnect = () => {
  if (shuttingDown || reconnectTimer) return;
  log("🔁 Reconnecting...", { delay: reconnectDelay });
  reconnectTimer = setTimeout(() => {
    reconnectTimer = null;
    connect();
    reconnectDelay = Math.min(reconnectDelay * 1.5, MAX_DELAY);
  }, reconnectDelay);
};

/*
---------------------------------------
SLEEP / WAKE DETECTION
---------------------------------------
*/

const SLEEP_CHECK_INTERVAL = 2000;
const SLEEP_THRESHOLD = SLEEP_CHECK_INTERVAL * 2;
let lastTickTime = Date.now();

setInterval(() => {
  const now = Date.now();
  const gap = now - lastTickTime;
  if (gap > SLEEP_THRESHOLD) {
    log(`😴 Wake detected (gap: ${gap}ms)`);
    immediateReconnect("System wake");
  }
  lastTickTime = now;
}, SLEEP_CHECK_INTERVAL);

/*
---------------------------------------
INTERNET RECOVERY DETECTION
Simply checks: do we have internet but no connection?
No state flags — works reliably every time.
---------------------------------------
*/

setInterval(() => {
  if (isConnected()) return;

  // Skip during cooldown window
  const timeSinceConnect = Date.now() - lastConnectedAt;
  if (timeSinceConnect < CONNECT_COOLDOWN) return;

  // Skip if already attempting or a retry is scheduled
  if (isConnecting || reconnectTimer) return;

  dns.lookup("1.1.1.1", (err) => {
    if (!err) {
      // Internet is up but we have no connection — reconnect immediately
      immediateReconnect("Internet available, not connected");
    }
  });
}, 2000);

/*
---------------------------------------
SHUTDOWN
---------------------------------------
*/

process.on("SIGINT", () => { shuttingDown = true; process.exit(0); });
process.on("SIGTERM", () => { shuttingDown = true; process.exit(0); });

/*
---------------------------------------
START
---------------------------------------
*/

connect();

File: backend/agent/agent.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net"
	"os"
	"path/filepath"
	"sync"
	"syscall"
	"os/signal"
	"time"

	"github.com/docker/docker/api/types"
	"github.com/docker/docker/api/types/container"
	"github.com/docker/docker/client"
	"github.com/gorilla/websocket"
)

const (
	WS_URL          = "wss://dockmon.onrender.com/agent"
	GITHUB_REPO     = "Hargun-singhh/DockMon"
	CURRENT_VERSION = "1.0.5"
	PING_INTERVAL   = 15 * time.Second
	PONG_TIMEOUT    = 8 * time.Second
	MAX_BACKOFF     = 5 * time.Second
	CONNECT_COOLDOWN = 4 * time.Second
)

type Agent struct {
	mu              sync.Mutex
	conn            *websocket.Conn
	docker          *client.Client
	deviceToken     string
	reconnectDelay  time.Duration
	lastConnectedAt time.Time
	isConnecting    bool
	shuttingDown    bool
}

func log(msg string) {
	fmt.Printf("[%s] %s\n", time.Now().UTC().Format(time.RFC3339), msg)
}

func logf(format string, args ...interface{}) {
	log(fmt.Sprintf(format, args...))
}

/*
---------------------------------------
AUTO UPDATE
---------------------------------------
*/

type GithubRelease struct {
	TagName string `json:"tag_name"`
	Assets  []struct {
		Name               string `json:"name"`
		BrowserDownloadURL string `json:"browser_download_url"`
	} `json:"assets"`
}

func checkAndUpdate() {
	return
}

/*
---------------------------------------
TOKEN
---------------------------------------
*/

func getToken() string {
	home, _ := os.UserHomeDir()
	configPath := filepath.Join(home, ".dockmon", "config.json")
	data, err := os.ReadFile(configPath)
	if err != nil {
		return ""
	}
	var config map[string]string
	if err := json.Unmarshal(data, &config); err != nil {
		return ""
	}
	return config["deviceToken"]
}

/*
---------------------------------------
DOCKER HANDLERS
---------------------------------------
*/

func (a *Agent) handleCommand(command string, payload map[string]interface{}) interface{} {
	ctx := context.Background()

	switch command {
	case "list_containers":
		containers, err := a.docker.ContainerList(ctx, container.ListOptions{All: true})
		if err != nil {
			return map[string]string{"error": err.Error()}
		}
		result := make([]map[string]interface{}, len(containers))
		for i, c := range containers {
			result[i] = map[string]interface{}{
				"id": c.ID, "names": c.Names,
				"image": c.Image, "state": c.State, "status": c.Status,
			}
		}
		return result

	case "start_container":
		id, _ := payload["container_id"].(string)
		if err := a.docker.ContainerStart(ctx, id, container.StartOptions{}); err != nil {
			return map[string]string{"error": err.Error()}
		}
		return map[string]bool{"success": true}

	case "stop_container":
		id, _ := payload["container_id"].(string)
		if err := a.docker.ContainerStop(ctx, id, container.StopOptions{}); err != nil {
			return map[string]string{"error": err.Error()}
		}
		return map[string]bool{"success": true}

	case "restart_container":
		id, _ := payload["container_id"].(string)
		if err := a.docker.ContainerRestart(ctx, id, container.StopOptions{}); err != nil {
			return map[string]string{"error": err.Error()}
		}
		return map[string]bool{"success": true}

	case "remove_container":
		id, _ := payload["container_id"].(string)
		if err := a.docker.ContainerRemove(ctx, id, container.RemoveOptions{Force: true}); err != nil {
			return map[string]string{"error": err.Error()}
		}
		return map[string]bool{"success": true}

	case "logs":
		id, _ := payload["container_id"].(string)
		reader, err := a.docker.ContainerLogs(ctx, id, container.LogsOptions{
			ShowStdout: true, ShowStderr: true, Timestamps: true, Tail: "500",
		})
		if err != nil {
			return map[string]string{"error": err.Error()}
		}
		defer reader.Close()
		data, _ := io.ReadAll(reader)
		return string(data)

	case "stats":
		id, _ := payload["container_id"].(string)
		resp, err := a.docker.ContainerStats(ctx, id, false)
		if err != nil {
			return map[string]string{"error": err.Error()}
		}
		defer resp.Body.Close()
		var stats types.StatsJSON
		if err := json.NewDecoder(resp.Body).Decode(&stats); err != nil {
			return map[string]string{"error": err.Error()}
		}
		cpuDelta := float64(stats.CPUStats.CPUUsage.TotalUsage - stats.PreCPUStats.CPUUsage.TotalUsage)
		systemDelta := float64(stats.CPUStats.SystemUsage - stats.PreCPUStats.SystemUsage)
		cpuCount := float64(stats.CPUStats.OnlineCPUs)
		if cpuCount == 0 {
			cpuCount = float64(len(stats.CPUStats.CPUUsage.PercpuUsage))
		}
		cpu := 0.0
		if systemDelta > 0 && cpuDelta > 0 {
			cpu = (cpuDelta / systemDelta) * cpuCount * 100
		}
		return map[string]interface{}{
			"cpu_percent":  fmt.Sprintf("%.2f", cpu),
			"memory_usage": stats.MemoryStats.Usage,
			"memory_limit": stats.MemoryStats.Limit,
		}

	case "list_images":
		images, err := a.docker.ImageList(ctx, types.ImageListOptions{})
		if err != nil {
			return map[string]string{"error": err.Error()}
		}
		return images

	default:
		return map[string]string{"error": "Unsupported command"}
	}
}

/*
---------------------------------------
SEND
---------------------------------------
*/

func (a *Agent) send(v interface{}) error {
	a.mu.Lock()
	defer a.mu.Unlock()
	if a.conn == nil {
		return fmt.Errorf("not connected")
	}
	a.conn.SetWriteDeadline(time.Now().Add(10 * time.Second))
	return a.conn.WriteJSON(v)
}

/*
---------------------------------------
CONNECT
---------------------------------------
*/

func (a *Agent) connect() {
	a.mu.Lock()
	if a.shuttingDown || a.isConnecting {
		a.mu.Unlock()
		return
	}
	if a.conn != nil {
		a.conn.Close()
		a.conn = nil
	}
	a.isConnecting = true
	a.mu.Unlock()

	log("🔌 Connecting...")

	dialer := websocket.Dialer{HandshakeTimeout: 10 * time.Second}
	conn, _, err := dialer.Dial(WS_URL, nil)

	a.mu.Lock()
	a.isConnecting = false
	a.mu.Unlock()

	if err != nil {
		logf("⚠️ WS Error: %s", err)
		a.scheduleReconnect()
		return
	}

	a.mu.Lock()
	a.conn = conn
	a.lastConnectedAt = time.Now()
	a.reconnectDelay = time.Second
	a.mu.Unlock()

	log("✅ Connected")

	conn.WriteJSON(map[string]string{
		"type":         "register",
		"device_token": a.deviceToken,
	})

	// Heartbeat in background
	go a.heartbeat(conn)

	// Read loop
	for {
		var msg map[string]interface{}
		if err := conn.ReadJSON(&msg); err != nil {
			logf("⚠️ Read error: %s", err)
			break
		}

		msgType, _ := msg["type"].(string)
		if msgType == "ping" {
			continue
		}
		if msgType == "command" {
			go func(m map[string]interface{}) {
				requestID, _ := m["request_id"].(string)
				command, _ := m["command"].(string)
				payload, _ := m["payload"].(map[string]interface{})
				if payload == nil {
					payload = map[string]interface{}{}
				}
				logf("📥 Command {\"command\":\"%s\"}", command)
				result := a.handleCommand(command, payload)
				a.send(map[string]interface{}{
					"type": "response", "request_id": requestID, "data": result,
				})
			}(msg)
		}
	}

	a.mu.Lock()
	a.conn = nil
	timeSinceConnect := time.Since(a.lastConnectedAt)
	a.mu.Unlock()

	if timeSinceConnect < CONNECT_COOLDOWN {
		log("🔁 Brief disconnect — waiting 3s")
		time.Sleep(3 * time.Second)
		go a.connect()
	} else {
		a.scheduleReconnect()
	}
}

/*
---------------------------------------
HEARTBEAT
---------------------------------------
*/

func (a *Agent) heartbeat(conn *websocket.Conn) {
	ticker := time.NewTicker(PING_INTERVAL)
	defer ticker.Stop()

	conn.SetPongHandler(func(string) error {
		conn.SetReadDeadline(time.Now().Add(PING_INTERVAL + PONG_TIMEOUT))
		return nil
	})

	for range ticker.C {
		a.mu.Lock()
		current := a.conn
		a.mu.Unlock()
		if current != conn {
			return
		}

		conn.SetWriteDeadline(time.Now().Add(10 * time.Second))
		if err := conn.WriteMessage(websocket.PingMessage, nil); err != nil {
			logf("💔 Ping failed: %s", err)
			conn.Close()
			return
		}
		conn.SetReadDeadline(time.Now().Add(PONG_TIMEOUT))
	}
}

/*
---------------------------------------
RECONNECT
---------------------------------------
*/

func (a *Agent) scheduleReconnect() {
	a.mu.Lock()
	delay := a.reconnectDelay
	newDelay := a.reconnectDelay * 3 / 2
	if newDelay > MAX_BACKOFF {
		newDelay = MAX_BACKOFF
	}
	a.reconnectDelay = newDelay
	a.mu.Unlock()

	logf("🔁 Reconnecting... {\"delay\":%d}", delay.Milliseconds())
	time.Sleep(delay)
	go a.connect()
}

func (a *Agent) immediateReconnect(reason string) {
	a.mu.Lock()
	if a.shuttingDown || a.isConnecting || a.conn != nil {
		a.mu.Unlock()
		return
	}
	timeSinceConnect := time.Since(a.lastConnectedAt)
	a.mu.Unlock()

	if timeSinceConnect < CONNECT_COOLDOWN {
		return
	}

	logf("⚡ %s — reconnecting immediately", reason)
	go a.connect()
}

/*
---------------------------------------
SLEEP / WAKE DETECTION
---------------------------------------
*/

func (a *Agent) watchSleep() {
	ticker := time.NewTicker(2 * time.Second)
	defer ticker.Stop()
	last := time.Now()

	for t := range ticker.C {
		gap := t.Sub(last)
		if gap > 4*time.Second {
			logf("😴 Wake detected (gap: %dms)", gap.Milliseconds())
			a.mu.Lock()
			if a.conn != nil {
				a.conn.Close()
				a.conn = nil
			}
			a.mu.Unlock()
			a.immediateReconnect("System wake")
		}
		last = t
	}
}

/*
---------------------------------------
INTERNET RECOVERY DETECTION
Uses TCP dial to 1.1.1.1:443 — works even when DNS is broken
---------------------------------------
*/

func (a *Agent) watchInternet() {
	ticker := time.NewTicker(2 * time.Second)
	defer ticker.Stop()

	for range ticker.C {
		a.mu.Lock()
		connected := a.conn != nil
		timeSinceConnect := time.Since(a.lastConnectedAt)
		isConnecting := a.isConnecting
		a.mu.Unlock()

		if connected || isConnecting || timeSinceConnect < CONNECT_COOLDOWN {
			continue
		}

		// TCP dial to 1.1.1.1 — faster and more reliable than DNS lookup
		conn, err := net.DialTimeout("tcp", "1.1.1.1:443", 2*time.Second)
		if err == nil {
			conn.Close()
			a.immediateReconnect("Internet available, not connected")
		}
	}
}

/*
---------------------------------------
MAIN
---------------------------------------
*/

func main() {
	checkAndUpdate()

	token := getToken()
	if token == "" {
		fmt.Println("❌ No DEVICE_TOKEN found.")
		fmt.Println("👉 Run: dockmon-agent login")
		os.Exit(1)
	}

	dockerClient, err := client.NewClientWithOpts(
		client.FromEnv,
		client.WithAPIVersionNegotiation(),
	)
	if err != nil {
		fmt.Printf("❌ Docker error: %s\n", err)
		os.Exit(1)
	}
	defer dockerClient.Close()

	agent := &Agent{
		docker:         dockerClient,
		deviceToken:    token,
		reconnectDelay: time.Second,
	}

	go agent.connect()
	go agent.watchSleep()
	go agent.watchInternet()

	quit := make(chan os.Signal, 1)
	signal.Notify(quit, syscall.SIGINT, syscall.SIGTERM)
	<-quit

	agent.mu.Lock()
	agent.shuttingDown = true
	if agent.conn != nil {
		agent.conn.Close()
	}
	agent.mu.Unlock()

	log("👋 Agent stopped")
}
